File: astrum-game/src/main/java/com/novusforge/astrum/world/World.java
package com.novusforge.astrum.world;

import java.util.Map;
import java.util.function.Consumer;

public class World {
    private final ChunkManager chunkManager;
    private final int seed;

    public World(int seed) {
        this.seed = seed;
        this.chunkManager = new ChunkManager(seed);
    }

    public void update(float dt) {
        chunkManager.tick();
    }

    public void updatePlayerPosition(float x, float y, float z) {
        int cx = (int) Math.floor(x / Chunk.SIZE);
        int cy = (int) Math.floor(y / Chunk.SIZE);
        int cz = (int) Math.floor(z / Chunk.SIZE);
        chunkManager.updatePlayerPosition(cx, cy, cz);
    }

    public short getBlock(int worldX, int worldY, int worldZ) {
        return chunkManager.getBlock(worldX, worldY, worldZ);
    }

    public Chunk getChunk(int cx, int cy, int cz) {
        return chunkManager.getChunk(cx, cy, cz);
    }

    public ChunkMesh getMesh(int cx, int cy, int cz) {
        return chunkManager.getMesh(cx, cy, cz);
    }

    public Map<Long, ChunkMesh> getVisibleMeshes(float px, float py, float pz, float[] frustumPlanes) {
        int pcx = (int) Math.floor(px / Chunk.SIZE);
        int pcy = (int) Math.floor(py / Chunk.SIZE);
        int pcz = (int) Math.floor(pz / Chunk.SIZE);
        return chunkManager.getVisibleMeshes(pcx, pcy, pcz, frustumPlanes);
    }

    public boolean isChunkReady(int cx, int cy, int cz) {
        return chunkManager.isChunkReady(cx, cy, cz);
    }

    public int getPendingWorldGen() { return chunkManager.getPendingWorldGen(); }
    public int getPendingMeshGen() { return chunkManager.getPendingMeshGen(); }
    public int getLoadedChunkCount() { return chunkManager.getLoadedChunkCount(); }

    public ChunkManager getChunkManager() {
        return chunkManager;
    }

    public static void setBufferDeleter(Consumer<Long> deleter) {
        ChunkMesh.setBufferDeleter(deleter);
    }

    public void dispose() {
        chunkManager.dispose();
    }
}

File: src/main/kotlin/com/novusforge/astrum/world/World.kt
package com.novusforge.astrum.world

import java.util.function.BiConsumer

/**
 * World - Manages chunks and block access
 * Pre-Classic implementation
 * Kotlin conversion: Using primary constructor and idiomatic method structure.
 */
class World(val seed: Int = (System.currentTimeMillis() and 0xFFFFFFFFL).toInt()) {
    val chunkManager: ChunkManager = ChunkManager(seed)

    /**
     * Update player position for chunk loading
     */
    fun updatePlayerPosition(x: Float, y: Float, z: Float) {
        val cx = Math.floor((x / Chunk.SIZE).toDouble()).toInt()
        val cz = Math.floor((z / Chunk.SIZE).toDouble()).toInt()
        chunkManager.updatePlayerPosition(cx, 0, cz)
    }

    /**
     * Get block at world coordinates
     */
    fun getBlock(worldX: Int, worldY: Int, worldZ: Int): Short {
        return chunkManager.getBlock(worldX, worldY, worldZ)
    }

    /**
     * Set block at world coordinates
     */
    fun setBlock(worldX: Int, worldY: Int, worldZ: Int, blockId: Short) {
        val cx = Math.floorDiv(worldX, Chunk.SIZE)
        val cz = Math.floorDiv(worldZ, Chunk.SIZE)
        val lx = Math.floorMod(worldX, Chunk.SIZE)
        val ly = Math.floorMod(worldY, Chunk.SIZE)
        val lz = Math.floorMod(worldZ, Chunk.SIZE)

        val chunk = chunkManager.getChunk(cx, cz)
        if (chunk != null) {
            chunk.setBlock(lx, ly, lz, blockId)
            chunkManager.regenerateMesh(cx, 0, cz)
        }
    }

    /**
     * Get chunk at coordinates
     */
    fun getChunk(cx: Int, cz: Int): Chunk? {
        return chunkManager.getChunk(cx, cz)
    }

    /**
     * Get visible chunk meshes (with frustum culling)
     */
    fun getVisibleMeshes(px: Float, py: Float, pz: Float, frustumPlanes: FloatArray?): Map<Long, ChunkMesh> {
        val pcx = Math.floor((px / Chunk.SIZE).toDouble()).toInt()
        val pcz = Math.floor((pz / Chunk.SIZE).toDouble()).toInt()
        return chunkManager.getVisibleMeshes(pcx, 0, pcz, frustumPlanes)
    }

    val loadedChunkCount: Int get() = chunkManager.loadedChunkCount
    val pendingWorldGen: Int get() = chunkManager.pendingWorldGen
    val pendingMeshGen: Int get() = chunkManager.pendingMeshGen

    fun dispose() {
        chunkManager.dispose()
    }

    companion object {
        @JvmStatic
        fun setBufferDeleter(deleter: BiConsumer<Long, Long>) {
            ChunkMesh.setBufferDeleter(deleter)
        }
    }
}
